File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        int[] freq = new int[128];
        for(char ch: t.toCharArray()){
            freq[ch]++;
        }

        int sL = 0;
        int mL = Integer.MAX_VALUE;
        int count = 0;
        int left = 0;

        for(int right = 0; right < s.length(); right++){
            char r = s.charAt(right);
            if(freq[r] > 0){
                count++;
            }
            freq[r]--;
            while(count == t.length()){
                if(right - left + 1 < mL){
                    mL = right - left + 1;
                    sL = left;
                }

                char ch = s.charAt(left);
                freq[ch]++;
                if(freq[ch] > 0){
                    count--;
                }
                left++;
            }
        }
        return mL == Integer.MAX_VALUE ? "" : s.substring(sL, mL + sL);

    }
}

File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        
        # Frequency map for characters in t
        freq = Counter(t)
        
        left = 0
        count = 0
        min_len = float('inf')  # Equivalent to Integer.MAX_VALUE
        s_index = 0
        
        for right in range(len(s)):
            ch = s[right]
            
            # If the character is needed in the current window
            if freq[ch] > 0:
                count += 1
                
            freq[ch] -= 1
            
            # Try to shrink the window from the left
            while count == len(t):
                if right - left + 1 < min_len:
                    min_len = right - left + 1
                    s_index = left
                    
                left_char = s[left]
                freq[left_char] += 1
                
                # If a required character leaves the window fully
                if freq[left_char] > 0:
                    count -= 1
                    
                left += 1
                
        # Return empty string if no window found, otherwise return the substring slice
        return "" if min_len == float('inf') else s[s_index : s_index + min_len]
